errors.md

```diff
@@ -0,0 +1,156 @@
+
+_Note: It is possiable to only cache the `name` and `message` if the error could cross a boundary (based on the stack frames)._
+
+If _originalError_ does not have an `[[ErrorData]]` internal slot, then produce a generic message. E.g.:
+
+```js
+function x() {
+    throw someObject;
+}
+```
+
+It produces a TypeError with the following message:
+
+```
+wrapped function threw, error was uncaught exception: Object
+```
+
+_Note: FF implementation follows the logic described above._
+
+### New Error.Stacks
+
+A good error message is sometimes not enough. We encourage engines to enable their engine-specific stack introspection mechanisms for errors in ShadowRealms. In this case, the Error's stack is subject to censorship constraints.
+
+Some of the use-cases for ShadowRealm, e.g., the virtualization, require a mechanism to control the error's stack, so those accessing it should not observe that the program is running inside a ShadowRealm. Early versions of the spec were simply nulling out the stack when an error crosses the callable boundary, but that was not sufficient, in fact, it was faulty for two main reasons:
+
+1. It was impossible for developers to figure out where the error occurred, including errors from the module graph linkage phase.
+2. Errors originated from within a ShadowRealm, and observed inside the same ShadowRealm, were still leaking all the information.
+
+This document describes how to solve both of these problems by applying a censoring process for all errors.
+
+#### Censoring Error objects accessible within ShadowRealms
+
+The first step is to make sure that when an error of any kind is observed (by reference) inside a ShadowRealm instance, the host is censoring all stack frames that are not associated to the ShadowRealm instance itself. This allows the following example to work properly:
+
+```js
+try { null.foo } catch (e) {
+    console.log(e.stack);
+    throw e;
+}
+```
+
+In the example above, `e` never crossed a callable boundary, but regardless, it should not contain a stack frame associated to another realm. Assuming that this error is observed (accessed) in two different places, a try/catch from above when running inside the ShadowRealm instances, and another try/catch in the incubator realm, which happens to be the main page (top level window), we should see two different error's stack for the two different errors:
+
+1. ShadowRoot's `Error.stack`:
+```
+foo@http://127.0.0.1:8081/y.js:2:5
+bar@http://127.0.0.1:8081/y.js:6:12
+captureStack@http://127.0.0.1:8081/x.js:9:9
+```
+
+_Note: Only the 3 stack frames must be exposed on this error because they correspond to code evaluated inside the ShadowRealm associated to the error instance._
+
+2. Main page's `Error.stack`:
+```
+foo@http://127.0.0.1:8081/y.js:2:5
+bar@http://127.0.0.1:8081/y.js:6:12
+captureStack@http://127.0.0.1:8081/x.js:9:9
+runInSandbox@http://127.0.0.1:8081/demo.html:6:17
```

Comment:
nit but `runInSanbox` may be distracting here, and a `runVirtualized` or something like that may be more appropriate.